src/array_api_stubs/_draft/array_object.py

```diff
@@ -343,9 +341,52 @@ def __dlpack__(
 
         Notes
         -----
+        Major DLPack versions represent ABI breaks, minor versions represent
+        ABI-compatible additions (e.g., new enum values for new data types or
+        device types).
+
+        The ``max_version`` keyword was introduced in v2023.12, and goes
+        together with the ``DLManagedTensorVersioned`` struct added in DLPack
+        1.0. This keyword may not be used by consumers for some time after
+        introduction. It is recommended to use this logic in the implementation
+        of ``__dlpack__``:
+
+        .. code:: python
+
+            if max_version is None:
+                # Keep and use the DLPack 0.X implementation
+                # Note: from March 2025 onwards (but ideally as late as
+                # possible), it's okay to raise BufferError here
+            else:
+                # We get to produce `DLManagedTensorVersioned` now
+                if max_version >= our_own_dlpack_version:
+                    # Consumer understands us, just return a Capsule with our max version
+                elif max_version[0] == our_own_dlpack_version[0]:
+                    # major versions match, we should still be fine here -
+                    # return our own max version
+                else:
+                    # if we're at a higher major version internally, did we
+                    # keep an implementation of the older major version around?
+                    # If so, use that. Else, just return our max
+                    # version and let the consumer deal with it.
```

Comment:
I see, so if a consumer pass `max_version=(0, 8)`, and if the producer is on 1.x, it is up to the producer to decide what to return.

Reply:
Revisiting this after some time, I think I was wrong. `max_version` is coming from the consumer, so if we're at this branch, it means the consumer likely cannot understand the versioned struct, and the producer can just raise `BufferError`. Please a look at the changes in commit 1db1717 and let me know if this is acceptable.

Reply:
Yes, I think this is right - and especially I think we can know it is right for `0.x` vs. `1.x`, since the actual struct changed. There are some ABI differences where it should be okay to return a newer version than requested, but there's also not much to gain there. So always raising `BufferError` here looks good to me.

Resolving the comment.

Reply:
Just to note, it currently says "should" which I think is correct:
* the producer knows it cannot really work (for major version being higher), so raising is better.
* The consumer still must check, we don't rely on the producer behavior here.

(I think this is fine, because consumers really must/should do some version checking anyway to cast to the right ABI struct.)